quarkus/runtime/src/test/java/or/keycloak/quarkus/runtime/cli/PicocliTest.java

```diff
@@ -255,4 +260,112 @@ public void spiRuntimeWarnWithBuild() {
+        assertEquals(CommandLine.ExitCode.SOFTWARE, nonRunningPicocli.exitCode);
+    }
+    
+    @Test
+    public void failStartBuildDev() {
+        NonRunningPicocli nonRunningPicocli = pseudoLaunch("--profile=dev", "start");
+        assertThat(nonRunningPicocli.getErrString(), containsString("You can not 'start' the server in development mode."));
+        assertEquals(CommandLine.ExitCode.SOFTWARE, nonRunningPicocli.exitCode);
+    }
+    
+    @Test
+    public void failIfOptimizedUsedForFirstStartupExport() {
+        NonRunningPicocli nonRunningPicocli = pseudoLaunch("export", "--optimized", "--dir=data");
+        assertEquals(CommandLine.ExitCode.USAGE, nonRunningPicocli.exitCode);
+        assertThat(nonRunningPicocli.getErrString(), containsString("The '--optimized' flag was used for first ever server start."));
+    }
+    
+    @Test
+    public void testReaugFromProdToDev() {
+        build("build");
+
+        Environment.setRebuildCheck(); // will be reset by the system properties logic
+        NonRunningPicocli nonRunningPicocli = pseudoLaunch("start-dev", "--hostname=name", "--http-enabled=true");
+        assertEquals(CommandLine.ExitCode.OK, nonRunningPicocli.exitCode);
+        assertTrue(nonRunningPicocli.reaug);
+        assertEquals("dev", nonRunningPicocli.buildProps.getProperty(org.keycloak.common.util.Environment.PROFILE));;
+    }
+
+    /**
+     * Runs a fake build to setup the state of the persisted build properties
+     */
+    private void build(String... args) {
+        if (Stream.of(args).anyMatch("start-dev"::equals)) {
+            Environment.setRebuildCheck(); // auto-build
+        }
+        NonRunningPicocli nonRunningPicocli = pseudoLaunch(args);
+        assertTrue(nonRunningPicocli.reaug);
+        assertEquals(CommandLine.ExitCode.OK, nonRunningPicocli.exitCode);
+        onAfter();
+        addPersistedConfigValues((Map)nonRunningPicocli.buildProps);
+    }
+    
+    @Test
+    public void testReaugFromProdToDevExport() {
+        build("build");
+        
+        Environment.setRebuildCheck(); // will be reset by the system properties logic
+        NonRunningPicocli nonRunningPicocli = pseudoLaunch("--profile=dev", "export", "--file=file");
+        assertEquals(CommandLine.ExitCode.OK, nonRunningPicocli.exitCode);
+        assertTrue(nonRunningPicocli.reaug);
+    }
+    
+    @Test
+    public void testNoReaugFromProdToExport() {
+        build("build");
+        
+        Environment.setRebuildCheck(); // will be reset by the system properties logic
+        NonRunningPicocli nonRunningPicocli = pseudoLaunch("export", "--file=file");
+        assertEquals(CommandLine.ExitCode.OK, nonRunningPicocli.exitCode);
+        assertFalse(nonRunningPicocli.reaug);
+    }
+    
+    @Test
+    public void testReaugFromDevToProd() {
+        build("start-dev");
+        
+        Environment.setRebuildCheck(); // will be reset by the system properties logic
```

Comment:
The `Environment.setRebuildCheck()` might be part of the `build(..)` method, right? Maybe considering it for further follow-ups.

Reply:
It could be, but it probably should be an option. We have three distinct test paths:
build + rebuildCheck = reaug test
build + w/o rebuildCheck = auto start test
build + w/o rebuildCheck + --optimized = optimized start test

I'm not sure if it's worth the effort, but we could change the rebuild check logic - rather than requiring 2 separate java commands we could just always do the rebuild check if --optimized is not supplied, then return a non-zero exit code when rebuilt to let the script no that another launch is needed.

Reply:
I see. Let's discuss some improvements around PicocliTest in follow-up tasks, how we could improve devEx.